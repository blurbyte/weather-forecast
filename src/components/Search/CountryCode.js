import styled from 'styled-components';

const CountryCode = styled.span`
  color: rgba(34, 34, 34, 0.5);
  text-transform: uppercase;
  margin-left: 0.4rem;
`;

export default CountryCode;
