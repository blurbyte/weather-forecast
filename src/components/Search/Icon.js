import styled from 'styled-components';

const Wrapper = styled.div`
  position: absolute;
  top: 0.8rem;
  right: 1rem;
  transition: color 0.35s linear;
`;

export default Wrapper;
