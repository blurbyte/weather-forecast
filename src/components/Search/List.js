import styled from 'styled-components';

const List = styled.ul`
  position: absolute;
  top: 3.9rem;
  right: 0;
  width: 36rem;
  border: 1px solid #222222;
  background-color: #ffffff;
  margin: 0;
  padding: 0;
  cursor: pointer;
`;

export default List;
