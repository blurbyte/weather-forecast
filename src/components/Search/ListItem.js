import styled from 'styled-components';

const List = styled.li`
  display: flex;
  align-items: center;
  height: 4rem;

  &:hover {
    background-color: rgba(34, 34, 34, 0.1);
  }
`;

export default List;
