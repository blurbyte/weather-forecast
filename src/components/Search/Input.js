import styled from 'styled-components';

import BaseInput from '../Input';
import Icon from './Icon';

const Input = styled(BaseInput)`
  padding-right: 4.4rem;

  & + ${Icon} {
    color: ${props => props.theme.primary};
  }

  &:focus + ${Icon} {
    color: #222;
  }
`;

export default Input;
