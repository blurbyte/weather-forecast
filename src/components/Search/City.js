import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ListItem from './ListItem';
import CountryCode from './CountryCode';

const Caption = styled.p`
  padding: 0 1.2rem;
  margin: 0;
`;

const City = ({ cityId, name, country, fetchForecast }) => (
  <ListItem onMouseDown={() => fetchForecast(cityId)}>
    <Caption>
      {name} <CountryCode>{country}</CountryCode>
    </Caption>
  </ListItem>
);

City.propTypes = {
  cityId: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  country: PropTypes.string.isRequired,
  fetchForecast: PropTypes.func.isRequired
};

export default City;
