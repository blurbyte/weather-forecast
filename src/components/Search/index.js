import React, { Component } from 'react';
import debounce from 'lodash.debounce';

import { getCities } from '../../services/forecast';
import { delay } from '../../utils';
import { Magnifier } from '../Icons';
import Input from './Input';
import Form from './Form';
import Icon from './Icon';

class Search extends Component {
  state = {
    search: '',
    foundCities: [],
    loading: false
  };

  fetchCities = async search => {
    this.setState({ loading: true });
    const cities = await getCities(search);
    await delay(1000);
    this.setState({ foundCities: cities, loading: false });
  };

  debounceFetch = debounce(() => {
    this.fetchCities(this.state.search);
  }, 400);

  handleSearch = e => {
    const search = e.target.value;
    // Functional state to trigger debounce after state change
    this.setState({ search }, () => {
      this.debounceFetch();
    });
  };

  render() {
    return (
      <Form>
        <Input type="text" placeholder="Your city name" onChange={this.handleSearch} />
        <Icon>
          <Magnifier />
        </Icon>
      </Form>
    );
  }
}

export default Search;
