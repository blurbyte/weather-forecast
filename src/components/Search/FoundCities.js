import React from 'react';
import PropTypes from 'prop-types';

import List from './List';
import City from './City';

const FoundCities = ({ cities }) => (
  <List>{cities.map(city => <City key={city.cityId} name={city.name} country={city.country} />)}</List>
);

FoundCities.propTypes = {
  cities: PropTypes.arrayOf(
    PropTypes.shape({
      cityId: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      country: PropTypes.string.isRequired
    })
  )
};

export default FoundCities;
