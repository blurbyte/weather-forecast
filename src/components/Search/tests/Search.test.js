import React from 'react';
import renderer from 'react-test-renderer';

import Search from '../index';

jest.mock('../../SearchInput', () => 'SearchInput');
jest.mock('../../Icons', () => ({
  Magnifier: 'Magnifier'
}));

const mockedEvent = {
  preventDefault: jest.fn()
};

const foundCities = [
  {
    cityId: 234567,
    name: 'London',
    country: 'US'
  },
  {
    cityId: 87654,
    name: 'London',
    country: 'GB'
  }
];

test('renders correctly', () => {
  const component = renderer.create(<Search fetchForecast={jest.fn()} />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('renders correctly when some cities got found', () => {
  const component = renderer.create(<Search fetchForecast={jest.fn()} />);

  const instance = component.getInstance();
  instance.setState({ foundCities });

  expect(component.toJSON()).toMatchSnapshot();
});

test('renders correctly when no matches got found', () => {
  const component = renderer.create(<Search fetchForecast={jest.fn()} />);

  const instance = component.getInstance();
  instance.setState({ foundCities: [], nothingFound: true });

  expect(component.toJSON()).toMatchSnapshot();
});

test('handleSubmit() handles submitting form correctly when some cities got found', () => {
  const fetchForecast = jest.fn();
  const component = renderer.create(<Search fetchForecast={fetchForecast} />);

  const instance = component.getInstance();
  instance.setState({ foundCities });
  instance.handleSubmit(mockedEvent);

  expect(fetchForecast).toHaveBeenCalledWith(foundCities[0].cityId);
});

test('handleSubmit() resets form if no cities were found', () => {
  const fetchForecast = jest.fn();
  const component = renderer.create(<Search fetchForecast={fetchForecast} />);

  const expectedState = {
    search: '',
    foundCities: [],
    nothingFound: false
  };

  const instance = component.getInstance();
  instance.setState({ foundCities: [], search: 'Taylorswift', nothingFound: true });
  instance.handleSubmit(mockedEvent);

  expect(instance.state).toEqual(expectedState);
});
