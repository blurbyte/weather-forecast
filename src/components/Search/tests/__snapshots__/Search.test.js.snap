// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="sc-bdVaJa iJQmjv"
  onSubmit={[Function]}
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Your city name"
    value=""
  />
</form>
`;

exports[`renders correctly when no matches got found 1`] = `
<form
  className="sc-bdVaJa iJQmjv"
  onSubmit={[Function]}
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Your city name"
    value=""
  />
  <div
    className="sc-bZQynM cgWLaR sc-EHOje kbodPI"
    onMouseDown={[Function]}
  >
    No cities found
  </div>
</form>
`;

exports[`renders correctly when some cities got found 1`] = `
<form
  className="sc-bdVaJa iJQmjv"
  onSubmit={[Function]}
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Your city name"
    value=""
  />
  <ul
    className="sc-bwzfXH ctBVmg"
  >
    <li
      className="sc-htpNat hQCJrQ"
      onMouseDown={[Function]}
    >
      <p
        className="sc-ifAKCX bnWgmO"
      >
        London
         
        <span
          className="sc-bxivhb bQvplE"
        >
          US
        </span>
      </p>
    </li>
    <li
      className="sc-htpNat hQCJrQ"
      onMouseDown={[Function]}
    >
      <p
        className="sc-ifAKCX bnWgmO"
      >
        London
         
        <span
          className="sc-bxivhb bQvplE"
        >
          GB
        </span>
      </p>
    </li>
  </ul>
</form>
`;
